### ISSUE TITLE
[TreeView] Add virtualization support

### ISSUE THREAD

**Issue by rashdeva** [mui/mui-x#9685]
Tree View doesn't support virtualization for long list items. It will be nice to have this feature. 

## Summary 💡
To maintain consistency with Autocomplete, it's better to use react-window as a virtualization library. To keep the current API, the Tree View component should merge items to flat list and by using indents present items. (https://material-ui.com/components/lists/#nested-list)

## Motivation 🔦
Most of the time tree view uses for long lists and structured data. If the parent item has 1000 children, the performance slows down.

**Comment by MpStyle** [comment]
Waiting for an official implementation, I attempted to implement a tree view with virtualization using react-virtualized-auto-sizer and react-window.
The data model is flat, and the depth is determined by a property (depth).
The React components are Material-UI components.
If anyone is interested, you can find it here:
[src/component/common/tree-view](https://github.com/MpStyle/wilde/tree/main/src/component/common/tree-view)

**Comment by sauron918** [comment]
Since the virtualization support is already on the [MUI X public roadmap](https://github.com/orgs/mui/projects/35) with a label `Plan: Pro`, should we expect that virtualization will be available for the Pro version only?

**Comment by oliviertassinari** [comment]
@sauron918 This is the current plan, yes. We are trying to make this effort sustainable, per our research (see issue description benchmark), the range of options available is very slim. See https://www.reddit.com/r/reactjs/comments/16vt1vv/react_component_needed_virtual_trees_w_drag_and/ for an interesting thread on this. I think this was borderline a Premium feature (close to no open-source alternatives) but Pro feels less greedy.

**Comment by itayG98** [comment]
Hello
I just wanted to ask if there any estimation to publish data for the virtual-tree at the pro plan? :)

**Comment by flaviendelangle** [comment]
We are aiming for an initial release of the virtualization in `@mui/x-tree-viewy-pro` for the launch of the version 8.0.0, probably in March.
The feature might be ready sooner as part of the v8 beta.

**Comment by itayG98** [comment]
> @itayG98 We are aiming for an initial release of the virtualization in `@mui/x-tree-view-pro` for the launch of MUI X v8.0.0, probably in March 2025. The feature might be ready sooner as part of the v8 beta.

Great ! I wish you luck!

**Comment by YassineElbouchaibi** [comment]
Hello! Thank you @flaviendelangle for the work you have been providing on this feature! Do you have an idea if this is still on track to maybe be introduced in March 2025? If you don't think so, do you have another estimate to share?

Thank you very much for all!

**Comment by flaviendelangle** [comment]
Hi,

We have some issues around this feature, we won't be able to launch it in March, but hopefully we should be able to resume the work on it soon.

**Comment by httpete-broadcom** [comment]
We are very interested in this. We have to stick with Arborist just for the virtualization.

**Comment by roland-rocely** [comment]
I built myself a virtualization hook for the TreeView. In my case, the app is wrapped in a Box with an ID of `ROOT_APP_CONTAINER_ID` - You might want to change the `document.querySelector` to be the document body.

`useVirtualization.ts`
```TypeScript
import { ROOT_APP_CONTAINER_ID } from "@/app/[language]/RootLayoutClientProviders";
import { useState, useEffect, useCallback } from "react";

export const useVirtualization = <T extends HTMLElement>(): [((node: T | null) => void), boolean] => {
    const [node, setNode] = useState<T | null>(null);
    const [isVisible, setIsVisible] = useState<boolean>(false);
    
    // This callback ref will be called every time the node changes.
    const ref = useCallback((node: T | null) => {
        setNode(node);
    }, []);
    
    useEffect(() => {
        if (!node) return;
        
        const container = document.querySelector(`#${ROOT_APP_CONTAINER_ID}`);
        const observer = new IntersectionObserver(
            (entries) => {
                entries.forEach((entry) => {
                    setIsVisible(entry.isIntersecting);
                });
            },
            {
                root: container,
                threshold: 0,
                rootMargin: "100px",
            }
        );
        
        observer.observe(node);
        
        return () => {
            observer.disconnect();
        };
    }, [node]); // re-run this effect any time the node changes
    
    return [ref, isVisible];
};
```

`DataTreeItem.tsx`
```TypeScript
export const DataTreeItem = forwardRef(function CustomTreeItem(
    props: CustomTreeItemProps,
    ref: Ref<HTMLLIElement>,
) {
    // ...
    const {
        children,
        // ...
    } = props;

    const [virtualRef, isVisible] = useVirtualization<HTMLLIElement>();
    
    // Merge the forwarded ref with the virtualization ref.
    const setRefs = mergeRefs(ref, virtualRef);
    
    const {
        getContentProps,
        // ...
    } = useTreeItem2({
        // ...
        rootRef: setRefs,
    });

    return (
        <TreeItem2Provider itemId={itemId}>
            <TreeItem2Root {...getRootProps(other)} ref={setRefs}>
                <TreeItem2Content
                    {...getContentProps()}
                    sx={{ height: "51.42px" }}
                >
                    {isVisible && (
                        // Render virtualized content here, i.e. 
                    )}
                </TreeItem2Content>
                {children && <TreeItem2GroupTransition {...getGroupTransitionProps()} />}
            </TreeItem2Root>
        </TreeItem2Provider>
    );
};
```

**Comment by fellan1996** [comment]
We are also very interested in this feature. Any idea when it will be arriving?

**Comment by flaviendelangle** [comment]
The work on the virtualization has been halted due to other priorities.
But we should resume it soon and release it in one of the v8 minor releases.

**Comment by SamuelTrew** [comment]
@flaviendelangle Given there has been a good number of minor releases for v8 now, I was wondering if there was an updated timeline for virtualisation support?

Thank you!

**Comment by httpete-broadcom** [comment]
Stuck with Arborist which is dying - adding this would be  boom to those who need a high performance tree.

**Comment by liv-farsight** [comment]
about to implement a tree view, would love to use MUI instead of arborist since i need virtualization. any update here?

**Comment by ymulenll** [comment]
This is a very important functionality for a high performance tree, I hope we can have some progress soon, we really need it.

**Comment by rocelyblooming** [comment]
This solution is working fabulously for ~10.000 rather complex tree items

> I built myself a virtualization hook for the TreeView. In my case, the app is wrapped in a Box with an ID of `ROOT_APP_CONTAINER_ID` - You might want to change the `document.querySelector` to be the document body.
> 
> `useVirtualization.ts`
> 
> import { ROOT_APP_CONTAINER_ID } from "@/app/[language]/RootLayoutClientProviders";
> import { useState, useEffect, useCallback } from "react";
> 
> export const useVirtualization = <T extends HTMLElement>(): [((node: T | null) => void), boolean] => {
>     const [node, setNode] = useState<T | null>(null);
>     const [isVisible, setIsVisible] = useState<boolean>(false);
>     
>     // This callback ref will be called every time the node changes.
>     const ref = useCallback((node: T | null) => {
>         setNode(node);
>     }, []);
>     
>     useEffect(() => {
>         if (!node) return;
>         
>         const container = document.querySelector(`#${ROOT_APP_CONTAINER_ID}`);
>         const observer = new IntersectionObserver(
>             (entries) => {
>                 entries.forEach((entry) => {
>                     setIsVisible(entry.isIntersecting);
>                 });
>             },
>             {
>                 root: container,
>                 threshold: 0,
>                 rootMargin: "100px",
>             }
>         );
>         
>         observer.observe(node);
>         
>         return () => {
>             observer.disconnect();
>         };
>     }, [node]); // re-run this effect any time the node changes
>     
>     return [ref, isVisible];
> };
> `DataTreeItem.tsx`
> 
> export const DataTreeItem = forwardRef(function CustomTreeItem(
>     props: CustomTreeItemProps,
>     ref: Ref<HTMLLIElement>,
> ) {
>     // ...
>     const {
>         children,
>         // ...
>     } = props;
> 
>     const [virtualRef, isVisible] = useVirtualization<HTMLLIElement>();
>     
>     // Merge the forwarded ref with the virtualization ref.
>     const setRefs = mergeRefs(ref, virtualRef);
>     
>     const {
>         getContentProps,
>         // ...
>     } = useTreeItem2({
>         // ...
>         rootRef: setRefs,
>     });
> 
>     return (
>         <TreeItem2Provider itemId={itemId}>
>             <TreeItem2Root {...getRootProps(other)} ref={setRefs}>
>                 <TreeItem2Content
>                     {...getContentProps()}
>                     sx={{ height: "51.42px" }}
>                 >
>                     {isVisible && (
>                         // Render virtualized content here, i.e. 
>                     )}
>                 </TreeItem2Content>
>                 {children && <TreeItem2GroupTransition {...getGroupTransitionProps()} />}
>             </TreeItem2Root>
>         </TreeItem2Provider>
>     );
> };

**Comment by trevancrowe** [comment]
Hoping to see this feature added soon as well. Currently having issues with large trees and would prefer to stick with MUI if possible.

**Comment by fellan1996** [comment]
@flaviendelangle How is it going. Any timeline you can give us? Are you currently working on it?

**Comment by flaviendelangle** [comment]
Hi,

We are currently building a shared virtualization engine with the Data Grid product, to make sure we share all the performance improvements between our products.
Once this new virtualization engine is ready, the Tree View is one of the product that should benefit from it. We don't have a clear timeline yet.
Until then the trick proposed earlier should give a good performance boost if your dataset is not very large.

**Comment by lemmylem** [comment]
This is something we desperately need. The MUI tree is great. But the fact that it does not support virtualization kind of stalled our progress in getting this to customers.

The above hack really doesn't go a great job, either. It uses an intersection observer for each item in the tree and that's a lot of overhead. Also, there are UI artifacts from it in larger trees, like weird scrolling behavior, "jumps" and sometimes flashes. Not really a solution for production code.

**Comment by caillou** [comment]
@flaviendelangle Sorry for pinging you on this matter.

I have the following question: Do you know if this feature is still on some kind of roadmap? Could we expect it within a version or two?

I am not asking for an ETA, just wondering what your feeling is on this one. Wondering if I should accept the rare cases where the performance is sub-par, in the hope of a potential change down the line, or if I should consider implementing our own virtualised tree. Which seems like a monumentous undertaking, seeing how feature complete `RichTreeViewPro` is.

And as always: Thanks for all the work and attention to detail that goes into this component. It's been a bliss working with it!

Edit: I just saw that the ticket is actually `in progress`. Sorry for the unnecessary ping.

<img width="385" height="115" alt="Image" src="https://github.com/user-attachments/assets/f661a118-6f59-43a0-8983-dd586b8acfc4" />

**Comment by flaviendelangle** [comment]
Yes it is still planned
We had a lot of delayed during the development of a shared virtualization engine across all of our products.
Things are now advancing but I can't give any roadmap yet.

Sorry for the delay 👍

**Comment by InsidersByte** [comment]
@flaviendelangle do you know when this will be released? The docs don't seem to mention this and neither (as far as I could tell) did the latest releases of the package

**Comment by flaviendelangle** [comment]
We should releae the first alpha for the next major next week I think.
And this will contain the tree view virtualization 👍

**Comment by itayG98** [comment]
Are there any plans to add virtualization capabillity to SimpleTreeView? I want to use my own component to render all tree items (same component for each), but need virtualization support for performance with large data sets.
Is this supported or on the roadmap?

**Comment by flaviendelangle** [comment]
No it's not planned.
What use case do you have that require using the Simple Tree View ? You can provide custom items to the Rich Tree View as well

**Comment by itayG98** [comment]
I noticed in the documentation at [https://mui.com/x/react-tree-view/rich-tree-view/items/]() that it explicitly states:

_Unlike SimpleTreeView, RichTreeView only supports string labels. You cannot pass React nodes to it_

Based on this, I understand that passing custom React components as tree items is not supported in RichTreeView.
Could you clarify if there’s a recommended approach for achieving custom rendering within RichTreeView items? I need to render my own item component with custom logic and styling.

**Comment by flaviendelangle** [comment]
We could probably make that doc more explicit.
You have examples [here](https://mui.com/x/react-tree-view/rich-tree-view/customization/) on how to customize the rendering.
For RIch Tree View and Simple Tree View, we recommand to create a custom item component if you have deep customization rather than passing a complex label (that could break features like the type ahead).

**Comment by justinmwarner** [comment]
What version of Tree List is/will this change be available? We're looking forward to getting this integrated! =)

**Comment by flaviendelangle** [comment]
It is available in the v9 alpha release (beta should start soon).
It was a breaking change because it required some DOM changes, but the migration should be very straightforward since wse don't have lots of other breaking changes.
